### ISSUE TITLE
_dumbwin32proc breaks with unicode environmental variables

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3964]
|<img alt="ivank's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| ivank reported|
|-|-|
|Trac ID|trac#3964|
|Type|defect|
|Created|2009-08-15 03:48:32Z|
|Branch|https://github.com/twisted/twisted/tree/unicode-environ-3964-2|

Windows supports unicode environmental variables. pywin32 supports spawning processes with str or unicode environmental variables, but not a mix of them:

```
...
  File "C:\Python26\lib\site-packages\twisted\internet\utils.py", line 25, in _callProtocolWithDeferred
    reactor.spawnProcess(p, executable, (executable,)+tuple(args), env, path)
  File "C:\Python26\lib\site-packages\twisted\internet\posixbase.py", line 234, in spawnProcess
    return Process(self, processProtocol, executable, args, env, path)
  File "C:\Python26\lib\site-packages\twisted\internet\_dumbwin32proc.py", line 176, in __init__
    doCreate()
  File "C:\Python26\lib\site-packages\twisted\internet\_dumbwin32proc.py", line 174, in doCreate
    command, cmdline, None, None, 1, 0, env, path, StartupInfo)
exceptions.TypeError: All dictionary items must be strings, or all must be unicode
```

This problem and "solution" is sort of documented here:
https://realityforge.vrsource.org/viewvc/maestro/trunk/maestro/daemon/plugins/services/launch/process.py , near "win32process.CreateProcess requires that all environment keys and"

This bug causes a lot of test failures with twisted.web2 on Windows Vista or Windows 7. The default install of Python 2.6.2 on these operating systems introduces environmental variables with unicode values:

C:\>python -c "import Tkinter, os; exec 'for k, v in os.environ.iteritems():\n\tif isinstance(v, unicode): print k, v'"
TK_LIBRARY C:\Python26\tcl\tk8.5
TIX_LIBRARY C:\Python26\tcl\tix8.4.3
TCL_LIBRARY C:\Python26\tcl\tcl8.5

<details><summary>Searchable metadata</summary>

```
trac-id__3964 3964
type__defect defect
reporter__ivank ivank
priority__highest highest
milestone__ 
branch__branches_unicode_environ_3964_2 branches/unicode-environ-3964-2
branch_author__pahan pahan
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1250308112000000 1250308112000000
changetime__1300129917000000 1300129917000000
version__None None
owner__PenguinOfDoom PenguinOfDoom
cc__thijs
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="ivank's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_1"></a>|ivank commented|
|-|-|

Both keys and values are unicode, or they are both ANSI, because of how CreateProcess works:

http://msdn.microsoft.com/en-us/library/ms682425%28VS.85%29.aspx

"
lpEnvironment

[...]

An environment block consists of a null-terminated block of null-terminated strings. Each string is in the following form:

name=value\0

Because the equal sign is used as a separator, it must not be used in the name of an environment variable.

An environment block can contain either Unicode or ANSI characters. If the environment block pointed to by lpEnvironment contains Unicode characters, be sure that dwCreationFlags includes CREATE_UNICODE_ENVIRONMENT.

[...]
"

**Comment by twisted-trac** [comment]
|<img alt="ivank's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_2"></a>|ivank commented|
|-|-|

setting this to high priority, given that it very often breaks `spawnProcess` on Vista/Win7

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph set owner to PenguinOfDoom|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="pahan's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_4"></a>|pahan commented|
|-|-|

(In [28748]) Branching to 'unicode-environ-3964'

**Comment by twisted-trac** [comment]
|<img alt="PenguinOfDoom's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|PenguinOfDoom removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_6"></a>|@jesstess set owner to PenguinOfDoom|
|-|-|

There don't appear to be commits on this branch. Removing from review.

Also, the link in the ticket description sounds very valuable but 404s for me.

**Comment by twisted-trac** [comment]
|<img alt="PenguinOfDoom's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_7"></a>|PenguinOfDoom removed owner|
|-|-|

oops :((((

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)|@jesstess set owner to @jesstess|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_9"></a>|@jesstess set owner to PenguinOfDoom|
|-|-|

Thanks for tackling this, PenguinOfDoom! Some feedback:

test_process.py
 * use the plural forms of asserts in test_encodableUnicodeEnvironment
 * use 2 blank lines between functions, as per the [coding standard](http://twistedmatrix.com/documents/current/core/development/policy/coding-standard.html)
 * Given the changes in _dumbwin23proc.py, I would have expected a test case that tries to set up both string and unicode environment variables and asserted that no errors were raised and they were all converted to unicode on process creation, which doesn't seem to be what's going on in test_encodableUnicodeEnvironment. I don't have a Windows machine on which I can play around with this test case easily, so I could be wrong about what's going on - can you clarify what's going on in the test's docstring if it is handling this case and/or add this scenario as a new test case?

_dumbwin23proc.py
 *  needs a copyright bump
 * has an unused failure import
 *  can you clarify this comment: "# in year 2000, win32process.CreateProcess cannot deal with mixed str/unicode environment, so we make it all Unicode"? Or maybe just take out the "in year 2000" part since it's a problem even now in 2010.
 * Process.__init__ is doing an insane amount compared to your average __init__, so it really needs a docstring, in conjunction with epytext markup for the instance variables in the class docstring. doCreate could also use a docstring.
 * There's a TODO above doCreate. If it is still relevant, it needs a ticket.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun commented|
|-|-|

(In [29358]) Fix some simple coding standard issues and update to a more modern testing style

refs [#3964](https://github.com/twisted/twisted/issues/3964)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun commented|
|-|-|

(In [29359]) Remove a confusing joke from this comment; meager beginning to a docstring for `Process.__init__`

refs [#3964](https://github.com/twisted/twisted/issues/3964)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun commented|
|-|-|

(In [29360]) Point to some existing tickets

refs [#3964](https://github.com/twisted/twisted/issues/3964)

**Comment by twisted-trac** [comment]
|<img alt="pahan's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_13"></a>|pahan commented|
|-|-|

(In [30958]) Branching to 'unicode-environ-3964-2'

**Comment by twisted-trac** [comment]
|<img alt="PenguinOfDoom's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|PenguinOfDoom removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_16"></a>|@exarkun set owner to PenguinOfDoom<br>@exarkun set status to `new`|
|-|-|

Build results look okay.  Unfortunate about that string comparison in the exception handler... Perhaps we can file a bug against pywin32 to get a custom exception type to detect this case?  Otherwise, fine.  Add a news file describing the bug fix and please merge.

**Comment by twisted-trac** [comment]
|<img alt="pahan's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_17"></a>|pahan set status to `closed`|
|-|-|

(In [31033]) Merge unicode-environ-3964-2: Support Unicode in system environment under Windows

Author: PenguinOfDoom
Reviewer: exarkun
Fixes: [#3964](https://github.com/twisted/twisted/issues/3964)

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_18"></a>|@thijstriemstra commented|
|-|-|

Replying to [pahan](#note_17):
> Reviewer: exarkun

aww jesstess :(
